### PR TITLE
storage,kvserver: Foundational changes for disaggregated ingestions

### PR DESCRIPTION
**Pull request by itsbilal**
This change contains two commits (split off from the original mega-PR, #105839). The first is a pkg/storage change to add new interface methods to call pebble's db.ScanInternal as well as implement related helper methods in sstable writers/batch readers/writers to be able to do disaggregated snapshot ingestion. The second is a kvserver/rditer change to allow finer-grained control on what replicated spans we iterate on, as well as to be able to specifically opt into skip-shared iteration over the user key span through the use of `ScanInternal`.

---

**storage: Update Engine/Reader/Writer interfaces for ScanInternal**

This change updates pkg/storage interfaces and implementations to allow
the use of ScanInternal in skip-shared iteration mode as well as
writing/reading of internal point keys, range dels and range keys.
Replication / snapshot code will soon rely on these changes to
be able to replicate internal keys in higher levels plus metadata
of shared sstables in lower levels, as opposed to just observed
user keys.

Part of https://github.com/cockroachdb/cockroach/issues/103028

Epic: none

Release note: None

**kvserver: Add ability to filter replicated spans in Select/Iterate**

This change adds the ability to select for just the replicated
span in rditer.Select and rditer.IterateReplicaKeySpans. Also
adds a new rditer.IterateReplicaKeySpansShared that does a
ScanInternal on just the user key span, to be able to collect
metadata of shared sstables as well as any internal keys above
them.

We only use skip-shared iteration for the replicated user key span
of a range, and in practice, only if it's a non-system range.

Part of https://github.com/cockroachdb/cockroach/issues/103028.

Epic: none

Release note: None

### CODE REVIEW

**Comment by itsbilal** [comment]
This is two smaller parts plucked in from #105839 and split into separate commits, to ease code review.

**Review by jbowens — COMMENTED** [review]
> Reviewed 5 of 8 files at r1, 4 of 49 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @itsbilal, and @pavelkalinnikov)

___
*[`pkg/storage/engine.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-N_u743h6cKZkEKG0nZ9:-N_u743h6cKZkEKG0nZA:b-r8rsum) ([raw file](https://github.com/cockroachdb/cockroach/blob/6862371071c097edaae1d7d83521927d2c1b6f99/pkg/storage/engine.go#L786)):*
> ```Go
> 	//
> 	// It is safe to modify the contents of the arguments after it returns.
> 	PutInternalKey(key *pebble.InternalKey, value []byte) error
> ```

Do these need to be a part of the `Writer` interface? Would it be feasible to extract them out into an `InternalWriter` that implements `Writer` and expands the interface? Since we don't expect any other callers to make use of these facilities, it'd be nice to quarantine them. Same goes for `ScanInternal` on `Reader`.

**Review by itsbilal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @jbowens, and @pavelkalinnikov)

___
*[`pkg/storage/engine.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-N_u743h6cKZkEKG0nZ9:-Na7cp4tBUUU2cZYpFo9:b-xap2dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6862371071c097edaae1d7d83521927d2c1b6f99/pkg/storage/engine.go#L786)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Do these need to be a part of the `Writer` interface? Would it be feasible to extract them out into an `InternalWriter` that implements `Writer` and expands the interface? Since we don't expect any other callers to make use of these facilities, it'd be nice to quarantine them. Same goes for `ScanInternal` on `Reader`.
</blockquote></details>

Done. The `InternalWriter` stuff turned out to be pretty straightforward as we only use these methods in the SST writer (where we directly reference the implementation), and the `WriteBatch` which is easy enough to use with `InternalWriter`. The Reader side is more complex as 1) Pebble supports ScanInternal on both snapshots and DBs, 2) we use it with a Snapshot, 3) `pebbleReadOnly` should probably also support `ScanInternal` anyway, 4) the only case where we don't support a ScanInternal on a reader is a batch, and we can just continue to panic there. So I left the Reader side as-is with the ScanInternal addition

**Review by bananabrick — COMMENTED** [review]
Will take another look at this tomorrow.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-:-Na8XPUp-uVowmGJ1AW6:b-bnc68w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @pavelkalinnikov)

**Review by bananabrick — COMMENTED** [review]
The code looks good to me. Haven't looked at all of the test files yet. Definitely deserves another look from someone familiar with the kvserver code.

> Reviewed 3 of 8 files at r1, 11 of 66 files at r3, 12 of 61 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-:-Nacr_X15m6e7jKHnEWo:b-e497lq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, and @pavelkalinnikov)

___
*[`pkg/storage/engine.go` line 878 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NachMf400_HCaXCRqSX:-NachMf400_HCaXCRqSY:bb4sk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/95a413eaa10719f3ad231d828eac3b3142aee5ed/pkg/storage/engine.go#L878)):*
> ```Go
> 	// start, end keys and bypasses the EngineKey encoding step. It also only
> 	// operates on point keys; for range keys, use ClearEngineRangeKey or
> 	// PutInternalRangeKey.
> ```

Is it not safe to modify contents of the slices after this function returns?
___
*[`pkg/storage/pebble_batch.go` line 577 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Nacjs6z15yIGGnXK7PL:-Nacjs6z15yIGGnXK7PM:b-nhccxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/95a413eaa10719f3ad231d828eac3b3142aee5ed/pkg/storage/pebble_batch.go#L577)):*
> ```Go
> }
> 
> // PutInternalKey implements the WriteBatch interface.
> ```

Is this still the `WriteBatch` interface?
___
*[`pkg/storage/pebble_batch.go` line 579 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NacjuuhBu-5O9TwxbQo:-NacjuuhBu-5O9TwxbQp:b-p4kiqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/95a413eaa10719f3ad231d828eac3b3142aee5ed/pkg/storage/pebble_batch.go#L579)):*
> ```Go
> // PutInternalKey implements the WriteBatch interface.
> func (p *pebbleBatch) PutInternalKey(key *pebble.InternalKey, value []byte) error {
> 	if len(key.UserKey) == 0 {
> ```

When does this happen?
___
*[`pkg/storage/sst_writer.go` line 274 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NackB160jbbQfdnwtB1:-NackB160jbbQfdnwtB2:bz1titz) ([raw file](https://github.com/cockroachdb/cockroach/blob/95a413eaa10719f3ad231d828eac3b3142aee5ed/pkg/storage/sst_writer.go#L274)):*
> ```Go
> }
> 
> // PutInternalKey implements the Writer interface.
> ```

Should this be `... InternalWriter interface`?
___
*[`pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/all/output` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Nacnk-e2FfusbSaE5P1:-Nacnk-e2FfusbSaE5P2:b-ek1cct) ([raw file](https://github.com/cockroachdb/cockroach/blob/422c94e36b19ccdbbcc467305877cb2cdb47935a/pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/all/output#L1)):*
> ```
> echo
> ```

This test is very hard to read. All these lines are long enough that they don't fit on a screen, and it looks like the text itself is wrapped? My editor has line wrapping turned off. Is it the same for you?

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/all/output` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Nacnk-e2FfusbSaE5P1:-NacsCoN2GObbB4bMo2y:bbdjcn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/422c94e36b19ccdbbcc467305877cb2cdb47935a/pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/all/output#L1)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

This test is very hard to read. All these lines are long enough that they don't fit on a screen, and it looks like the text itself is wrapped? My editor has line wrapping turned off. Is it the same for you?
</blockquote></details>

I was able to fix this. Was a problem with my editor.

**Review by itsbilal — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-:-Nald7xq5uWE3GKv-Kxk:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @jbowens, and @pavelkalinnikov)

___
*[`pkg/storage/engine.go` line 878 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NachMf400_HCaXCRqSX:-Nald6-39aW3eRW2zCBH:b-owxgfm) ([raw file](https://github.com/cockroachdb/cockroach/blob/95a413eaa10719f3ad231d828eac3b3142aee5ed/pkg/storage/engine.go#L878)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Is it not safe to modify contents of the slices after this function returns?
</blockquote></details>

Clarified. And yes.
___
*[`pkg/storage/pebble_batch.go` line 577 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Nacjs6z15yIGGnXK7PL:-Nalcn74AvtPwidZt0rS:b8ckjtn) ([raw file](https://github.com/cockroachdb/cockroach/blob/95a413eaa10719f3ad231d828eac3b3142aee5ed/pkg/storage/pebble_batch.go#L577)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Is this still the `WriteBatch` interface?
</blockquote></details>

Yes, because WriteBatch composes InternalBatch.
___
*[`pkg/storage/pebble_batch.go` line 579 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NacjuuhBu-5O9TwxbQo:-NalcozZ9eGNgfwtD8kl:bq5463a) ([raw file](https://github.com/cockroachdb/cockroach/blob/95a413eaa10719f3ad231d828eac3b3142aee5ed/pkg/storage/pebble_batch.go#L579)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

When does this happen?
</blockquote></details>

The same check is in all other Batch methods too, so it's here for consistency.
___
*[`pkg/storage/sst_writer.go` line 274 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NackB160jbbQfdnwtB1:-NalctbwCcRitg469-03:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/95a413eaa10719f3ad231d828eac3b3142aee5ed/pkg/storage/sst_writer.go#L274)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Should this be `... InternalWriter interface`?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/all/output` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Nacnk-e2FfusbSaE5P1:-NalcudE4I76XEoUkpVP:b-vkp2z3) ([raw file](https://github.com/cockroachdb/cockroach/blob/422c94e36b19ccdbbcc467305877cb2cdb47935a/pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/all/output#L1)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

I was able to fix this. Was a problem with my editor.
</blockquote></details>

It's largely autogenerated code, so I would mostly look at differences between this and the other files with the different replicated key filters, rather than review the whole thing as-is.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed 2 of 8 files at r1, 2 of 66 files at r3, 12 of 65 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @itsbilal, @jbowens, and @pavelkalinnikov)

___
*[`pkg/storage/engine.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-N_u743h6cKZkEKG0nZ9:-NarbaU36Nd6Ps3l9uR6:b-10bv18) ([raw file](https://github.com/cockroachdb/cockroach/blob/6862371071c097edaae1d7d83521927d2c1b6f99/pkg/storage/engine.go#L786)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Done. The `InternalWriter` stuff turned out to be pretty straightforward as we only use these methods in the SST writer (where we directly reference the implementation), and the `WriteBatch` which is easy enough to use with `InternalWriter`. The Reader side is more complex as 1) Pebble supports ScanInternal on both snapshots and DBs, 2) we use it with a Snapshot, 3) `pebbleReadOnly` should probably also support `ScanInternal` anyway, 4) the only case where we don't support a ScanInternal on a reader is a batch, and we can just continue to panic there. So I left the Reader side as-is with the ScanInternal addition
</blockquote></details>

In CRDB, we will be doing ScanInternal only on EFOS, yes?

I'm ok with it being defined on `Reader` if that is the simplest option.
___
*[`pkg/storage/engine.go` line 605 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Naran059BhoPktfBs61:-Naran059BhoPktfBs62:bexan4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/engine.go#L605)):*
> ```Go
> 	// InternalKeys using a visitor pattern, while also allowing for keys in
> 	// shared files to be skipped if a visitor is provided for visitSharedFiles.
> 	// Useful for fast-replicating state from one Reader to another.
> ```

`DB.ScanInternal` allows `visitSharedFile` to be nil, and I don't see a use case for that. Can we slightly simplify the Pebble code base by requiring it to be non-nil?
___
*[`pkg/storage/engine.go` line 892 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NarfNPv2RueRpPlf9bk:-NarfNPv2RueRpPlf9bl:b-ckc6t3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/engine.go#L892)):*
> ```Go
> 	//
> 	// It is safe to modify the contents of the arguments after it returns.
> 	PutInternalKey(key *pebble.InternalKey, value []byte) error
> ```

PutInternalPointKey?
___
*[`pkg/storage/open.go` line 187 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NarfwsI3ow4CkcrYhn1:-NarfwsI3ow4CkcrYhn2:b-wx6185) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/open.go#L187)):*
> ```Go
> 		// version upgrade finalization. However, seeing as shared storage is
> 		// an experimental feature and upgrading from existing stores is not
> 		// supported, this is fine.
> ```

This will ratchet the Pebble format major version, so if someone starts such a node and then wants to start an older binary it will fail, yes? I suppose that can't be made to work anyway since we could have introduced remote files.
btw, have we finalized on the terminology in https://docs.google.com/document/d/1BnjKvE0RERV0QkTAfR5iNn8sYVo0gaVgDx4x6kCNCGg/edit and if yes, is there a place in the Pebble code which documents it?
___
*[`pkg/storage/pebble.go` line 2478 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Narhs5T5Wbj50EXnbVO:-Narhs5T5Wbj50EXnbVP:b-9by05t) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/pebble.go#L2478)):*
> ```Go
> 	visitSharedFile func(sst *pebble.SharedSSTMeta) error,
> ) error {
> 	return p.parent.ScanInternal(ctx, lower, upper, visitPointKey, visitRangeDel, visitRangeKey, visitSharedFile)
> ```

Given this appeals to the parent, it does not satisfy the behavior of `ConsistentIterators`. We should document in `Reader.ScanInternal` that it is not covered by the guarantees of `ConsistentIterators`.
___
*[`pkg/storage/pebble_batch.go` line 275 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Nare3C3D7b4GIYWdThI:-Nare3C3D7b4GIYWdThJ:b-tdoq6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/pebble_batch.go#L275)):*
> ```Go
> // ClearRawEncodedRange implements the InternalWriter interface.
> func (p *pebbleBatch) ClearRawEncodedRange(start, end []byte) error {
> 	return p.batch.DeleteRange(start, end, pebble.Sync)
> ```

why `pebble.Sync`? I was expecting nil since this is a batch.
___
*[`pkg/storage/pebble_batch.go` line 504 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NareYsk1SxmtY_quBHL:-NareYsk1SxmtY_quBHM:bjgyhvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/pebble_batch.go#L504)):*
> ```Go
> 
> // PutInternalRangeKey implements the InternalWriter interface.
> func (p *pebbleBatch) PutInternalRangeKey(start, end []byte, key rangekey.Key) error {
> ```

We expose the seqnums from `ScanInternal`, but don't use them.

The code comment for `DB.ScanInternal` does not say that it is collapsing range keys, but it must be doing so -- can you update that comment. Also, similar comment for `Reader.ScanInternal`.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed 5 of 61 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @itsbilal, @jbowens, and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/rditer/replica_data_iter.go` line 388 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NaroKEI7MjsaiQTTGdH:-NaroKEI7MjsaiQTTGdI:b-fcl9p6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a50a339323dbd15ce532d2d744770c58ab3317de/pkg/kv/kvserver/rditer/replica_data_iter.go#L388)):*
> ```Go
> 		panic("reader must provide consistent iterators")
> 	}
> 	span := desc.KeySpan().AsRawSpanWithNoLocals()
> ```

Should we use `Select` to construct this with `ReplicatedSpansUserOnly`? I don't see any other reason for the existence of `ReplicatedSpansUserOnly`.
It also gives us better test coverage -- I had to check that `desc.KeySpan()` does the forwarding of the start key to `LocalMax`.

**Review by itsbilal — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-:-Naw01KN3b4aQmqXksMU:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @jbowens, @pavelkalinnikov, and @sumeerbhola)

___
*[`pkg/kv/kvserver/rditer/replica_data_iter.go` line 388 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NaroKEI7MjsaiQTTGdH:-Naw-bdu6QPCf57ooXAZ:b-6wr7eh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a50a339323dbd15ce532d2d744770c58ab3317de/pkg/kv/kvserver/rditer/replica_data_iter.go#L388)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Should we use `Select` to construct this with `ReplicatedSpansUserOnly`? I don't see any other reason for the existence of `ReplicatedSpansUserOnly`.
It also gives us better test coverage -- I had to check that `desc.KeySpan()` does the forwarding of the start key to `LocalMax`.
</blockquote></details>

Done. The other use case of `ReplicatedSpansUserOnly` is when we restart iteration due to `ErrInvalidSkipSharedIteration`.
___
*[`pkg/storage/engine.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-N_u743h6cKZkEKG0nZ9:-Nas5c143lBYxJo2sHB7:bix26wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6862371071c097edaae1d7d83521927d2c1b6f99/pkg/storage/engine.go#L786)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

In CRDB, we will be doing ScanInternal only on EFOS, yes?

I'm ok with it being defined on `Reader` if that is the simplest option.
</blockquote></details>

Yeah, we'll probably only use it on EFOS from CRDB. There's the ScanStatistics case but that's probably going to be within Pebble.
___
*[`pkg/storage/engine.go` line 605 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Naran059BhoPktfBs61:-Navx2nc9ZzACGE8FIdX:b3x4cjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/engine.go#L605)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

`DB.ScanInternal` allows `visitSharedFile` to be nil, and I don't see a use case for that. Can we slightly simplify the Pebble code base by requiring it to be non-nil?
</blockquote></details>

The nil use case is `ScanStatistics` and potentially others in the future. We can require it to be non-nil in the Cockroach interface though, if we want.
___
*[`pkg/storage/engine.go` line 892 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NarfNPv2RueRpPlf9bk:-Nas6ijEAGPzVt5-O-hO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/engine.go#L892)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

PutInternalPointKey?
</blockquote></details>

Done.
___
*[`pkg/storage/open.go` line 187 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NarfwsI3ow4CkcrYhn1:-Nas6m2j6burgWq-7r6h:b-y6se9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/open.go#L187)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This will ratchet the Pebble format major version, so if someone starts such a node and then wants to start an older binary it will fail, yes? I suppose that can't be made to work anyway since we could have introduced remote files.
btw, have we finalized on the terminology in https://docs.google.com/document/d/1BnjKvE0RERV0QkTAfR5iNn8sYVo0gaVgDx4x6kCNCGg/edit and if yes, is there a place in the Pebble code which documents it?
</blockquote></details>

I think this use of SharedStorage is accurate as it'll only be used for shared sstables, not all remote sstables. But yes, we've finalized the terminology and the best definition for the different terms is around BackingType in db.go in pebble. 
___
*[`pkg/storage/pebble.go` line 2478 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Narhs5T5Wbj50EXnbVO:-NavxJ9T6OxcPz0asYFn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/pebble.go#L2478)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Given this appeals to the parent, it does not satisfy the behavior of `ConsistentIterators`. We should document in `Reader.ScanInternal` that it is not covered by the guarantees of `ConsistentIterators`.
</blockquote></details>

Done.
___
*[`pkg/storage/pebble_batch.go` line 275 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-Nare3C3D7b4GIYWdThI:-NavyBdMBBR-trqKfsPQ:bi14ar1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/pebble_batch.go#L275)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why `pebble.Sync`? I was expecting nil since this is a batch.
</blockquote></details>

All the other use-cases in `pebbleBatch` pass in `pebble.Sync` for the opt, so I just maintained consistency. Either way it won't make a difference as we're not committing the batch.
___
*[`pkg/storage/pebble_batch.go` line 504 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-NareYsk1SxmtY_quBHL:-Naw-IaG7LCwfgP-rmJa:bc9xwc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62647735726f9e7353c81e292d1d4c2aaa5a87/pkg/storage/pebble_batch.go#L504)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We expose the seqnums from `ScanInternal`, but don't use them.

The code comment for `DB.ScanInternal` does not say that it is collapsing range keys, but it must be doing so -- can you update that comment. Also, similar comment for `Reader.ScanInternal`.
</blockquote></details>

Done. DB.ScanInternal in Pebble already talks about this , though it could be more specific when talking about range key collapsing.

File: pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/only/output
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/only/output:1]: What is "only" in this test name? Is this a leftover test after "only" was renamed to "user-only"? Or it's something else; but there is a "user-only" test too, so I'm not sure what the difference should be.

File: pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r3/replicatedOnly/skip/output
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r3/replicatedOnly/skip/output:1]: Trying to match this "skip" to the test code, but can't find what it corresponds too. Could you help me find it? Or it's a leftover file after some rename?

File: pkg/kv/kvserver/rditer/testdata/TestSelect/r1
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/rditer/testdata/TestSelect/r1:3]: It would be good to extend this test to have `ReplicatedSpansFilter` other than "include all".

**Review by pav-kv — COMMENTED** [review]
Reviewed the `rditer` commit. Looks good. Some questions about the tests, there seem to be some leftover unused data files.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: once @pavelkalinnikov is happy with the `rditer` tests

> Reviewed 1 of 61 files at r6, 5 of 66 files at r7, 5 of 61 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107297#-:-NbGddpe18BiXS8gx7XR:b-cel51q)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick, @itsbilal, @jbowens, and @pavelkalinnikov)

File: pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/exclude-user/output
Comment (itsbilal) [reply pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/exclude-user/output:23]: Yes, this is expected (and is the same as the previous behaviour of this test too). It's because we decode the lock table key and only pretty-print the MVCC portion of the key:

https://github.com/cockroachdb/cockroach/blob/4c397610e416cd81c839ec2f8b7cbc5ac51f8322/pkg/kv/kvserver/rditer/replica_data_iter_test.go#L186

File: pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/only/output
Comment (itsbilal) [reply pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/only/output:1]: You're right, this is an excess / unused file. Removed (along with others). Thanks!

File: pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r3/replicatedOnly/skip/output
Comment (itsbilal) [reply pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r3/replicatedOnly/skip/output:1]: Removed.

File: pkg/kv/kvserver/rditer/testdata/TestSelect/r1
Comment (itsbilal) [reply pkg/kv/kvserver/rditer/testdata/TestSelect/r1:3]: Done.

**Comment by itsbilal** [comment]
TFTRs! Addressed @pavelkalinnikov 's outstanding comments. Also rebased on #108369 to make it build which should be merged soon anyway (I'll rebase it on master once that merges)

**Pull request merged**
